erdjs/src/smartcontracts/typesystem/matchers.ts

```diff
@@ -0,0 +1,108 @@
+import * as errors from "../../errors";
+import { AddressType, AddressValue } from "./address";
+import { BooleanType, BooleanValue } from "./boolean";
+import { OptionalType, OptionalValue, Vector, VectorType } from "./generic";
+import { NumericalType, NumericalValue } from "./numerical";
+import { Structure, StructureType } from "./structure";
+import { PrimitiveType, PrimitiveValue, Type } from "./types";
+
+export function onTypeSelect<TResult>(type: Type, selectors: {
+    onOptional: () => TResult,
+    onVector: () => TResult,
+    onPrimitive: () => TResult,
+    onStructure: () => TResult,
+    onOther?: () => TResult
+}): TResult {
+    if (type instanceof OptionalType) {
```

Comment:
Should the `TypeRegistry` help here?

Reply:
It is possible to add the following on class `Type`:

```
/**
 * Inspired from: https://docs.microsoft.com/en-us/dotnet/api/system.type.isassignablefrom
 */
isAssignableFrom(type: Type): boolean 
    return type instanceof this.constructor;
}
```

And then do:

```
for (knownType of [BooleanType.One, NumericalType.One, AddressType.One]) {
	if (knownType.isAssignableFrom(type)) {
	    func = selectors[knownType.Name];
	    return func();
	}
}
```

But then I would loose the type safety and the compiler checks around the `selectors` parameter.

I've added though the `isAssignableFrom` function, in case we have some other ideas.